CP lock LockOwnershipLostException: Current thread is not owner of the Lock

http://jenkins.hazelcast.com/view/cp-system/job/cp-multiStruts-stable/1/console

/disk1/jenkins/workspace/cp-multiStruts-stable/3.12-SNAPSHOT/2019_02_19-03_00_49/multi-structs


fail HzMember1HZ cp_tryLock hzcmd.cp.lock.TryLocker threadId=0 com.hazelcast.cp.lock.exception.LockOwnershipLostException: Current thread is not owner of the Lock[cp_tryLock12] because its Session[1] is closed by server!


http://54.227.64.79/~jenkins/workspace/cp-multiStruts-stable/3.12-SNAPSHOT/2019_02_19-03_00_49/multi-structs/



this is a stable cluster test using all cp structs concurrently 



```
cat output/HZ/HzMember1HZ/exception.txt
com.hazelcast.cp.lock.exception.LockOwnershipLostException: Current thread is not owner of the Lock[cp_tryLock12] because its Session[1] is closed by server!
	at com.hazelcast.cp.internal.datastructures.lock.proxy.AbstractRaftFencedLockProxy.newLockOwnershipLostException(AbstractRaftFencedLockProxy.java:361)
	at com.hazelcast.cp.internal.datastructures.lock.proxy.AbstractRaftFencedLockProxy.unlock(AbstractRaftFencedLockProxy.java:219)
	at hzcmd.cp.lock.TryLocker.timeStep(TryLocker.java:23)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
java.lang.IllegalStateException: No valid session!
	at com.hazelcast.cp.internal.datastructures.semaphore.proxy.RaftSessionAwareSemaphoreProxy.newIllegalStateException(RaftSessionAwareSemaphoreProxy.java:246)
	at com.hazelcast.cp.internal.datastructures.semaphore.proxy.RaftSessionAwareSemaphoreProxy.release(RaftSessionAwareSemaphoreProxy.java:168)
	at com.hazelcast.cp.internal.datastructures.semaphore.proxy.RaftSessionAwareSemaphoreProxy.release(RaftSessionAwareSemaphoreProxy.java:150)
	at hzcmd.cp.semaphore.TryAcquire.timeStep(TryAcquire.java:26)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.cp.internal.session.SessionExpiredException: active session: 1 does not exist in CPGroupId{name='default', seed=0, commitIndex=8}
	at com.hazelcast.cp.internal.datastructures.spi.blocking.AbstractBlockingService.heartbeatSession(AbstractBlockingService.java:260)
	at com.hazelcast.cp.internal.datastructures.semaphore.RaftSemaphoreService.releasePermits(RaftSemaphoreService.java:102)
	at com.hazelcast.cp.internal.datastructures.semaphore.operation.ReleasePermitsOp.run(ReleasePermitsOp.java:53)
	at com.hazelcast.cp.internal.NodeEngineRaftIntegration.runOperation(NodeEngineRaftIntegration.java:184)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.applyLogEntry(RaftNodeImpl.java:601)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.applyLogEntries(RaftNodeImpl.java:551)
	at com.hazelcast.cp.internal.raft.impl.handler.AppendSuccessResponseHandlerTask.commitEntries(AppendSuccessResponseHandlerTask.java:162)
	at com.hazelcast.cp.internal.raft.impl.handler.AppendSuccessResponseHandlerTask.handleResponse(AppendSuccessResponseHandlerTask.java:88)
	at com.hazelcast.cp.internal.raft.impl.handler.AbstractResponseHandlerTask.innerRun(AbstractResponseHandlerTask.java:43)
	at com.hazelcast.cp.internal.raft.impl.task.RaftNodeStatusAwareTask.run(RaftNodeStatusAwareTask.java:46)
	at com.hazelcast.cp.internal.NodeEngineRaftIntegration.execute(NodeEngineRaftIntegration.java:97)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.execute(RaftNodeImpl.java:636)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.handleAppendResponse(RaftNodeImpl.java:239)
	at com.hazelcast.cp.internal.RaftService.handleAppendResponse(RaftService.java:605)
	at com.hazelcast.cp.internal.operation.integration.AppendSuccessResponseOp.run(AppendSuccessResponseOp.java:48)com.hazelcast.cp.lock.exception.LockOwnershipLostException: Current thread is not owner of the Lock[cp_tryLock12] because its Session[1] is closed by server!
	at com.hazelcast.spi.Operation.call(Operation.java:170)

	at com.hazelcast.cp.internal.datastructures.lock.proxy.AbstractRaftFencedLockProxy.newLockOwnershipLostException(AbstractRaftFencedLockProxy.java:361)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:210)	at com.hazelcast.cp.internal.datastructures.lock.proxy.AbstractRaftFencedLockProxy.unlock(AbstractRaftFencedLockProxy.java:219)

	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:199)	at hzcmd.cp.lock.TryLocker.timeStep(TryLocker.java:23)

	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:416)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:153)	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)

	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:123)	at remote.bench.BenchThread.call(BenchThread.java:41)

	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:110)	at remote.bench.BenchThread.call(BenchThread.java:14)

	at ------ submitted from ------.(Unknown Source)	at java.util.concurrent.FutureTask.run(FutureTask.java:266)

	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:126)	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)

	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:79)	at java.util.concurrent.FutureTask.run(FutureTask.java:266)

	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:162)	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:177)	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)

	at com.hazelcast.spi.impl.AbstractInvocationFuture.join(AbstractInvocationFuture.java:143)	at java.lang.Thread.run(Thread.java:748)

	at com.hazelcast.cp.internal.datastructures.semaphore.proxy.RaftSessionAwareSemaphoreProxy.release(RaftSessionAwareSemaphoreProxy.java:165)
	... 12 more
java.lang.IllegalStateException: No valid session!
	at com.hazelcast.cp.internal.datastructures.semaphore.proxy.RaftSessionAwareSemaphoreProxy.newIllegalStateException(RaftSessionAwareSemaphoreProxy.java:246)
	at com.hazelcast.cp.internal.datastructures.semaphore.proxy.RaftSessionAwareSemaphoreProxy.release(RaftSessionAwareSemaphoreProxy.java:168)
	at com.hazelcast.cp.internal.datastructures.semaphore.proxy.RaftSessionAwareSemaphoreProxy.release(RaftSessionAwareSemaphoreProxy.java:150)
	at hzcmd.cp.semaphore.TryAcquire.timeStep(TryAcquire.java:26)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.cp.internal.session.SessionExpiredException: active session: 1 does not exist in CPGroupId{name='default', seed=0, commitIndex=8}
	at com.hazelcast.cp.internal.datastructures.spi.blocking.AbstractBlockingService.heartbeatSession(AbstractBlockingService.java:260)
	at com.hazelcast.cp.internal.datastructures.semaphore.RaftSemaphoreService.releasePermits(RaftSemaphoreService.java:102)
	at com.hazelcast.cp.internal.datastructures.semaphore.operation.ReleasePermitsOp.run(ReleasePermitsOp.java:53)
	at com.hazelcast.cp.internal.NodeEngineRaftIntegration.runOperation(NodeEngineRaftIntegration.java:184)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.applyLogEntry(RaftNodeImpl.java:601)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.applyLogEntries(RaftNodeImpl.java:551)
	at com.hazelcast.cp.internal.raft.impl.handler.AppendSuccessResponseHandlerTask.commitEntries(AppendSuccessResponseHandlerTask.java:162)
	at com.hazelcast.cp.internal.raft.impl.handler.AppendSuccessResponseHandlerTask.handleResponse(AppendSuccessResponseHandlerTask.java:88)
	at com.hazelcast.cp.internal.raft.impl.handler.AbstractResponseHandlerTask.innerRun(AbstractResponseHandlerTask.java:43)
	at com.hazelcast.cp.internal.raft.impl.task.RaftNodeStatusAwareTask.run(RaftNodeStatusAwareTask.java:46)
	at com.hazelcast.cp.internal.NodeEngineRaftIntegration.execute(NodeEngineRaftIntegration.java:97)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.execute(RaftNodeImpl.java:636)
	at com.hazelcast.cp.internal.raft.impl.RaftNodeImpl.handleAppendResponse(RaftNodeImpl.java:239)
	at com.hazelcast.cp.internal.RaftService.handleAppendResponse(RaftService.java:605)
	at com.hazelcast.cp.internal.operation.integration.AppendSuccessResponseOp.run(AppendSuccessResponseOp.java:48)
	at com.hazelcast.spi.Operation.call(Operation.java:170)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:210)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:199)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:416)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:153)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:123)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:110)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:126)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:79)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:162)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:177)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.join(AbstractInvocationFuture.java:143)
	at com.hazelcast.cp.internal.datastructures.semaphore.proxy.RaftSessionAwareSemaphoreProxy.release(RaftSessionAwareSemaphoreProxy.java:165)
	... 12 more
```

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/handler/AppendRequestHandlerTask.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/handler/AppendRequestHandlerTask.java
@@ -127,6 +127,7 @@ public class AppendRequestHandlerTask extends RaftNodeStatusAwareTask implements
             }
         }
 
+        int truncatedAppendRequestEntryCount = 0;
         LogEntry[] newEntries = null;
         // Process any new entries
         if (req.entryCount() > 0) {
@@ -148,16 +149,17 @@ public class AppendRequestHandlerTask extends RaftNodeStatusAwareTask implements
                 // If an existing entry conflicts with a new one (same index but different terms),
                 // delete the existing entry and all that follow it (§5.3)
                 if (reqEntry.term() != localEntry.term()) {
-                    List<LogEntry> truncated = raftLog.truncateEntriesFrom(reqEntry.index());
+                    List<LogEntry> truncatedEntries = raftLog.truncateEntriesFrom(reqEntry.index());
                     if (logger.isFineEnabled()) {
-                        logger.warning("Truncated " + truncated.size() + " entries from entry index: " + reqEntry.index() + " => "
-                                + truncated);
+                        logger.warning("Truncated " + truncatedEntries.size() + " entries from entry index: "
+                                + reqEntry.index() + " => " + truncatedEntries);
                     } else {
-                        logger.warning("Truncated " + truncated.size() + " entries from entry index: " + reqEntry.index());
+                        logger.warning("Truncated " + truncatedEntries.size() + " entries from entry index: "
+                                + reqEntry.index());
                     }
 
                     raftNode.invalidateFuturesFrom(reqEntry.index());
-                    revertRaftGroupCmd(truncated);
+                    revertRaftGroupCmd(truncatedEntries);
 
                     newEntries = Arrays.copyOfRange(req.entries(), i, req.entryCount());
                     break;
@@ -165,6 +167,16 @@ public class AppendRequestHandlerTask extends RaftNodeStatusAwareTask implements
             }
 
             if (newEntries != null && newEntries.length > 0) {
+                if (raftLog.availableCapacity() < newEntries.length) {
+                    if (logger.isFineEnabled()) {
+                        logger.warning("Truncating " + newEntries.length + " entries to " + raftLog.availableCapacity()
+                                + " to fit into the available capacity of the Raft log");
+                    }
+
+                    truncatedAppendRequestEntryCount = newEntries.length - raftLog.availableCapacity();
+                    newEntries = Arrays.copyOf(newEntries, raftLog.availableCapacity());
+                }
+
                 // Append any new entries not already in the log
                 if (logger.isFineEnabled()) {
                     logger.fine("Appending " + newEntries.length + " entries: " + Arrays.toString(newEntries));
@@ -174,11 +186,14 @@ public class AppendRequestHandlerTask extends RaftNodeStatusAwareTask implements
             }
         }
 
-        long lastLogIndex = req.prevLogIndex() + req.entryCount();
+        // I cannot use raftLog.lastLogOrSnapshotIndex() for lastLogIndex because my log may contain
+        // some uncommitted entries from the previous leader and those entries will be truncated soon
+        // I can only send a response based on how many entries I have appended from this append request
+        long lastLogIndex = req.prevLogIndex() + req.entryCount() - truncatedAppendRequestEntryCount;
         long oldCommitIndex = state.commitIndex();
 
         // Update the commit index
-        if (req.leaderCommitIndex() > state.commitIndex()) {
+        if (req.leaderCommitIndex() > oldCommitIndex) {
             // If leaderCommit > commitIndex, set commitIndex = min(leaderCommit, index of last new entry)
             long newCommitIndex = min(req.leaderCommitIndex(), lastLogIndex);
             if (logger.isFineEnabled()) {

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/log/RaftLog.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/raft/impl/log/RaftLog.java
@@ -146,8 +146,18 @@ public class RaftLog {
         return truncated;
     }
 
+    /**
+     * Returns the number of empty indices in the Raft log
+     */
+    public int availableCapacity() {
+        return (int) (logs.getCapacity() - logs.size());
+    }
+
+    /**
+     * Returns true if the Raft log contains empty indices for the requested amount
+     */
     public boolean checkAvailableCapacity(int requestedCapacity) {
-        return (logs.getCapacity() - logs.size()) >= requestedCapacity;
+        return availableCapacity() >= requestedCapacity;
     }
 
     /**

--- a/hazelcast/src/test/java/com/hazelcast/cp/internal/raft/impl/LocalRaftTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cp/internal/raft/impl/LocalRaftTest.java
@@ -696,7 +696,7 @@ public class LocalRaftTest extends HazelcastTestSupport {
         leader.replicate(new ApplyRaftRunnable("val1")).get();
 
         final RaftNodeImpl[] followers = group.getNodesExcept(leader.getLocalMember());
-        final RaftNodeImpl followerWithLongerLog = followers[0];
+        final RaftNodeImpl followerWithLongestLog = followers[0];
         final long commitIndex = getCommitIndex(leader);
 
         assertTrueEventually(new AssertTask() {
@@ -717,7 +717,7 @@ public class LocalRaftTest extends HazelcastTestSupport {
         assertTrueEventually(new AssertTask() {
             @Override
             public void run() {
-                assertTrue(getLastLogOrSnapshotEntry(followerWithLongerLog).index() > commitIndex);
+                assertTrue(getLastLogOrSnapshotEntry(followerWithLongestLog).index() > commitIndex);
             }
         });
 
@@ -730,14 +730,15 @@ public class LocalRaftTest extends HazelcastTestSupport {
             }
         }, 10);
 
-        group.dropMessagesToMember(followerWithLongerLog.getLocalMember(), followers[1].getLocalMember(), VoteRequest.class);
-        group.dropMessagesToMember(followerWithLongerLog.getLocalMember(), followers[2].getLocalMember(), VoteRequest.class);
+        group.dropMessagesToMember(followerWithLongestLog.getLocalMember(), followers[1].getLocalMember(), VoteRequest.class);
+        group.dropMessagesToMember(followerWithLongestLog.getLocalMember(), followers[2].getLocalMember(), VoteRequest.class);
 
         group.terminateNode(leader.getLocalMember());
 
         RaftNodeImpl newLeader = group.waitUntilLeaderElected();
 
-        // followerWithLongerLog has 2 entries, other 3 followers have 1 entry and the 3 followers will elect their leaders
+        // followerWithLongestLog has 2 entries, other 3 followers have 1 entry
+        // and those 3 followers will elect a leader among themselves
 
         assertTrueEventually(new AssertTask() {
             @Override
@@ -745,7 +746,7 @@ public class LocalRaftTest extends HazelcastTestSupport {
                 for (RaftNodeImpl raftNode : followers) {
                     Endpoint newLeader = getLeaderMember(raftNode);
                     assertNotEquals(leader.getLocalMember(), newLeader);
-                    assertNotEquals(followerWithLongerLog.getLocalMember(), newLeader);
+                    assertNotEquals(followerWithLongestLog.getLocalMember(), newLeader);
                 }
             }
         });
@@ -755,17 +756,17 @@ public class LocalRaftTest extends HazelcastTestSupport {
             assertEquals(commitIndex, getLastLogOrSnapshotEntry(followers[i]).index());
         }
 
-        // followerWithLongerLog has not truncated its extra log entry until the new leader appends a new entry
-        assertTrue(getLastLogOrSnapshotEntry(followerWithLongerLog).index() > commitIndex);
+        // followerWithLongestLog does not truncate its extra log entry until the new leader appends a new entry
+        assertTrue(getLastLogOrSnapshotEntry(followerWithLongestLog).index() > commitIndex);
 
         newLeader.replicate(new ApplyRaftRunnable("val3")).get();
 
         assertTrueEventually(new AssertTask() {
             @Override
             public void run() {
-                for (RaftNodeImpl raftNode : followers) {
-                    assertEquals(2, getCommitIndex(raftNode));
-                    RaftDataService service = group.getService(raftNode);
+                for (RaftNodeImpl follower : followers) {
+                    assertEquals(2, getCommitIndex(follower));
+                    RaftDataService service = group.getService(follower);
                     assertEquals(2, service.size());
                     assertEquals("val1", service.get(1));
                     assertEquals("val3", service.get(2));
@@ -773,7 +774,7 @@ public class LocalRaftTest extends HazelcastTestSupport {
             }
         });
 
-        assertEquals(2, getLastLogOrSnapshotEntry(followerWithLongerLog).index());
+        assertEquals(2, getLastLogOrSnapshotEntry(followerWithLongestLog).index());
     }
 
     @Test

--- a/hazelcast/src/test/java/com/hazelcast/cp/internal/raft/impl/SnapshotTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cp/internal/raft/impl/SnapshotTest.java
@@ -627,4 +627,82 @@ public class SnapshotTest extends HazelcastTestSupport {
 
         assertFalse(f.isDone());
     }
+
+    @Test
+    public void when_slowFollowerReceivesAppendRequestThatDoesNotFitIntoItsRaftLog_then_itTruncatesAppendRequestEntries()
+            throws ExecutionException, InterruptedException {
+        int appendRequestMaxEntryCount = 100;
+        final int commitIndexAdvanceCount = 100;
+        int uncommittedEntryCount = 10;
+
+        RaftAlgorithmConfig config = new RaftAlgorithmConfig()
+                .setAppendRequestMaxEntryCount(appendRequestMaxEntryCount)
+                .setCommitIndexAdvanceCountToSnapshot(commitIndexAdvanceCount)
+                .setUncommittedEntryCountToRejectNewAppends(uncommittedEntryCount);
+        group = newGroupWithService(5, config);
+        group.start();
+
+        final RaftNodeImpl leader = group.waitUntilLeaderElected();
+        final RaftNodeImpl[] followers = group.getNodesExcept(leader.getLocalMember());
+        final RaftNodeImpl slowFollower1 = followers[0];
+        final RaftNodeImpl slowFollower2 = followers[1];
+
+        int count = 1;
+        for (int i = 0; i < commitIndexAdvanceCount; i++) {
+            leader.replicate(new ApplyRaftRunnable("val" + (count++))).get();
+        }
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                for (RaftNodeImpl node : group.getNodes()) {
+                    assertTrue(getSnapshotEntry(node).index() > 0);
+                }
+            }
+        });
+
+        group.dropMessagesToMember(leader.getLocalMember(), slowFollower1.getLocalMember(), AppendRequest.class);
+
+        for (int i = 0; i < commitIndexAdvanceCount - 1; i++) {
+            leader.replicate(new ApplyRaftRunnable("val" + (count++))).get();
+        }
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                assertEquals(getCommitIndex(leader), getCommitIndex(slowFollower2));
+            }
+        });
+
+        // slowFollower2's log: [ <91 - 100 before snapshot>, <100 snapshot>, <101 - 199 committed> ]
+
+        group.dropMessagesToMember(leader.getLocalMember(), slowFollower2.getLocalMember(), AppendRequest.class);
+
+        for (int i = 0; i < commitIndexAdvanceCount / 2; i++) {
+            leader.replicate(new ApplyRaftRunnable("val" + (count++))).get();
+        }
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                assertTrue(getSnapshotEntry(leader).index() > commitIndexAdvanceCount);
+            }
+        });
+
+        // leader's log: [ <191 - 199 before snapshot>, <200 snapshot>, <201 - 249 committed> ]
+
+        group.allowMessagesToMember(leader.getLocalMember(), slowFollower2.getLocalMember(), AppendRequest.class);
+
+        // leader replicates 50 entries to slowFollower2 but slowFollower2 has only available capacity of 11 indices.
+        // so, slowFollower2 appends 11 of these 50 entries in the first AppendRequest, takes a snapshot,
+        // and receives another AppendRequest for the remaining entries...
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                assertEquals(getCommitIndex(leader), getCommitIndex(slowFollower2));
+                assertTrue(getSnapshotEntry(slowFollower2).index() > commitIndexAdvanceCount);
+            }
+        });
+    }
 }

